File: C/Engine/Source/Runtime/Platform/Windows/FileDialog.c
#include "FileDialog.h"

#include "Windows.h"



typedef BOOL(APIENTRY*DialogCallback)(OPENFILENAMEW*);

static Bool OpenDialog(Char16* inBuffer, SizeType inSize, const Char16* filters, const Char16* initDirectory, const Char16* title, Void* owner, DialogCallback Fn)
{
	FMemory.Clear( inBuffer, sizeof(Char16) * inSize );

	OPENFILENAMEW openFilename;
	FMemory.Clear( &openFilename, sizeof(OPENFILENAMEW) );
	openFilename.lStructSize		= sizeof(OPENFILENAMEW);
	openFilename.hwndOwner			= CAST(HWND, owner);
	openFilename.lpstrFilter		= filters;
	openFilename.lpstrFile			= inBuffer;
	openFilename.nMaxFile			= inSize;
	openFilename.lpstrInitialDir	= initDirectory;
	openFilename.lpstrTitle			= title;
	openFilename.Flags				= OFN_NOCHANGEDIR | OFN_PATHMUSTEXIST | OFN_FILEMUSTEXIST;

	return Fn( &openFilename );
}


static Bool Open(Char16* inBuffer, SizeType inSize, const Char16* filters, const Char16* initDirectory, const Char16* title, Void* owner)
{
	return OpenDialog(inBuffer, inSize, filters, initDirectory, title, owner, GetOpenFileNameW);
}

static Bool Save(Char16* inBuffer, SizeType inSize, const Char16* filters, const Char16* initDirectory, const Char16* title, Void* owner)
{
	return OpenDialog(inBuffer, inSize, filters, initDirectory, title, owner, GetSaveFileNameW);
}


struct FFileDialog FFileDialog = 
{
	Open,
	Save,
};

File: CSharp/Engine/Runtime/Platform/Windows/FileDialog.cs
using System;
using System.Runtime.InteropServices;



#if PLATFORM_WINDOWS
using static Engine.Windows.APISet.CommonDialog;


namespace Engine.Windows
{
	public class FileDialog
	{
		public static bool Open(out string outputSelectedFile, string filter = "(*.*)\0*.*\0", string initDirectory = "./", string title = "Open file..", HWND owner = default)
		{
			return OpenDialog(out outputSelectedFile, filter, initDirectory, title, owner, GetOpenFileNameW);
		}

		public static bool Save(out string outputSelectedFile, string filter = "(*.*)\0*.*\0", string initDirectory = "./", string title = "Save as..", HWND owner = default)
		{
			return OpenDialog(out outputSelectedFile, filter, initDirectory, title, owner, GetSaveFileNameW);
		}


		private static bool OpenDialog(out string outputSelectedFile, string filter, string initDirectory, string title, HWND owner, Open Fn)
		{
			var ofn				= new OPENFILENAMEW();
			ofn.lStructSize		= (uint)Marshal.SizeOf<OPENFILENAMEW>();
			ofn.hwndOwner		= owner;
			ofn.lpstrFilter		= filter;
			ofn.lpstrFile		= new string( new char[256] );
			ofn.nMaxFile		= (uint)ofn.lpstrFile.Length;
			ofn.lpstrInitialDir	= initDirectory;
			ofn.lpstrTitle		= title;
			ofn.Flags			= OFN_NOCHANGEDIR | OFN_PATHMUSTEXIST | OFN_FILEMUSTEXIST;

			var errorCode = Fn(ref ofn);
			outputSelectedFile = ofn.lpstrFile;

			return errorCode != 0;
		}
	}
}
#endif
